velox/connectors/hive/HiveDataSource.cpp

```diff
@@ -599,8 +501,8 @@ void HiveDataSource::addSplit(std::shared_ptr<ConnectorSplit> split) {
 
   VLOG(1) << "Adding split " << split_->toString();
 
-  fileHandle_ = fileHandleFactory_->generate(split_->filePath).second;
-  auto input = createBufferedInput(*fileHandle_, readerOpts_);
+  auto input = createBufferedInput(
```

Comment:
`fileHandle_` must be kept to avoid eviction

Reply:
@Yuhta Could you please explain a bit? From what I saw, the fileHandle was moved into the BufferedInput anyways:
```
  BufferedInput(
      std::shared_ptr<ReadFile> readFile,
      memory::MemoryPool& pool,
      const MetricsLogPtr& metricsLog = MetricsLog::voidLog(),
      IoStatistics* FOLLY_NULLABLE stats = nullptr,
      uint64_t maxMergeDistance = kMaxMergeDistance,
      std::optional<bool> wsVRLoad = std::nullopt)
      : input_{std::make_shared<ReadFileInputStream>(
            std::move(readFile),
            metricsLog,
            stats)},
        pool_{pool},
        maxMergeDistance_{maxMergeDistance},
        wsVRLoad_{wsVRLoad},
        allocPool_{std::make_unique<memory::AllocationPool>(&pool)} {}
```

Reply:
The `FileHandle` is value from a LRU cache that could be evicted.  If you lose that, the next time you read the same file, it will not be able to retrieve the same file from memory and SSD cache, and degrade the performance.

Reply:
@Yuhta Thanks for the explanation. But as I understood, if a <filePath, fileHandle> entry was evicted from `fileHandleFactory_` (which is a CachedFactory), then `fileHandleFactory_->generate(split_->filePath)` would create a new fileHandle object and update `fileHandle_` anyways: 
```
  fileHandle_ = fileHandleFactory_->generate(split_->filePath).second;  
```
This `generate` function would create a different FileHandle object but with the same filePath and fileNum is as follows:
```
template <typename Key, typename Value, typename Generator>
std::pair<bool, Value> CachedFactory<Key, Value, Generator>::generate(
    const Key& key) {  // key is the filePath
  process::TraceContext trace("CachedFactory::generate");
  std::unique_lock<std::mutex> pending_lock(pendingMu_);
  {
    std::lock_guard<std::mutex> cache_lock(cacheMu_);
    auto value = cache_->get(key);     // Suppose the entry was evicted, the value won't be found
    if (value) {
      return std::make_pair(true, value.value());
    }
  }

  if (pending_.contains(key)) {
    pendingCv_.wait(pending_lock, [&]() { return !pending_.contains(key); });
    {
      std::lock_guard<std::mutex> cache_lock(cacheMu_);
      auto value = cache_->get(key);  // Suppose the entry was evicted, the value won't be found
      if (value) {
        return std::make_pair(true, value.value());
      }
    }
    pending_lock.unlock();
    return generate(key);    // Regenerate in the edge case.
  } else {
    pending_.insert(key);
    pending_lock.unlock();
    Value generatedValue;
    try {
      generatedValue = (*generator_)(key);    // Generate the fileHandle again
    } catch (const std::exception& e) {
      ...
    }
    ...
    return std::make_pair(false, generatedValue);
  }
}
```
And the generator is as follows
```
std::shared_ptr<FileHandle> FileHandleGenerator::operator()(
    const std::string& filename) {
  // We have seen cases where drivers are stuck when creating file handles.
  // Adding a trace here to spot this more easily in future.
  process::TraceContext trace("FileHandleGenerator::operator()");
  uint64_t elapsedTimeUs{0};
  std::shared_ptr<FileHandle> fileHandle;
  {
    MicrosecondTimer timer(&elapsedTimeUs);
    fileHandle = std::make_shared<FileHandle>();   // Create a new FileHandle object
    fileHandle->file = filesystems::getFileSystem(filename, properties_).  
                           ->openFileForRead(filename);   // will call open() and get a new ReadFile with a new fd_ member
    fileHandle->uuid = StringIdLease(fileIds(), filename);      // Create a StringIdLease for the same file, but the ids_ and id_ members for it would be the same. This is because the ids_ is a bi-map of the filename and ids, while not the fileHandle and ids.
    fileHandle->groupId = StringIdLease(fileIds(), groupName(filename)); // Create a StringIdLease for the same file, but the ids_ and id_ members for it would be the same.
  }
  ...
  return fileHandle;
}

```
So if the same file is encountered again but the previous fileHandle was evicted from `fileHandleFactory_`, the `fileHandle_` would be assigned to the newly generated fileHandle and no longer the previous object, but with the same uuid and groupId. Then, the cache_ uses `RawFileCacheKey` as the key to the files, which is based on the uuid:

```
struct RawFileCacheKey {
  uint64_t fileNum;
  uint64_t offset;
}
```
The CacheInputStream load data by checking if the fileNum, or uuid, is in the cache_
```
void CacheInputStream::loadSync(Region region) {
  cache::RawFileCacheKey key{fileNum_, region.offset};   // The fileNum_ is the UUID
  ...
  pin_ = cache_->findOrCreate(key, region.length, &wait);
}
```

So it seems to me it's not the FileHandle object that is the key to the cache, but the UUID for the fileName and region offset. As long as the new fileHandle contains the same UUID, there is no need to persist `fileHandle_` as a member of HiveDataSource. Plus, even if you persist it as a member, when `fileHandleFactory_->generate(split_->filePath)` is called again, the `fileHandle_` would still be re-created and updated. Is my understanding correct?

Reply:
@Yuhta I think it still makes sense to keep the FileHandleFactory. The generate() function may be expensive, especially the following line that opens a file:
```
fileHandle->file = filesystems::getFileSystem(filename, properties_).  
                           ->openFileForRead(filename);
```

Reply:
`fileNum` is `UUID.id()`. The `id()` is not guaranteed to be the same for the same file. It rotates.
```
  // Makes a lease for 'string' and makes sure it has an id.
  StringIdLease(StringIdMap& ids, std::string_view string)
      : ids_(&ids), id_(ids_->makeId(string)) {}
```
I also see Orri's comment here that https://github.com/prestodb/presto/issues/20019#issuecomment-1730489305 that HiveDataSouce is expected to own the FileHandle